syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

option go_package = ".;grpc_services";

service DataService {
  // User
  rpc LoginUser (UserLoginRequest) returns (google.protobuf.Empty) {}
  rpc CreateUser (UserCreateUpdateRequest) returns (User) {}
  rpc GetUser (UserReadDeleteRequest) returns (User) {}
  rpc UpdateUser (UserCreateUpdateRequest) returns (User) {}
  rpc DeleteUser (UserReadDeleteRequest) returns (google.protobuf.Empty) {}
  rpc ListUsers (google.protobuf.Empty) returns (Users) {}

  // Chat
  rpc CreateChat (ChatCreateRequest) returns (Chat) {}
  rpc GetChat (ChatReadRequest) returns (Chat) {}
  rpc UpdateChat (ChatUpdateRequest) returns (Chat) {}
  rpc DeleteChat (ChatDeleteRequest) returns (google.protobuf.Empty) {}
  rpc ListChats (UserReadDeleteRequest) returns (Chats) {}                      // Usa por conta do e-mail do usuario

  // Message
  rpc CreateMessage (MessageCreateRequest) returns (Message) {}
  rpc GetMessage (MessageReadRequest) returns (Message) {}
  rpc UpdateMessage (MessageUpdateRequest) returns (Message) {}
  rpc DeleteMessage (MessageDeleteRequest) returns (google.protobuf.Empty) {}
  rpc ListMessages (ChatDeleteRequest) returns (Messages) {}                    // Usa por conta do idChat
}

// ================== User =====================

message UserCreateUpdateRequest {
  string name = 1;
  string email = 2;
  string password = 3;
}

message UserReadDeleteRequest {
  string email = 1;
}

message UserLoginRequest {
  string email = 1;
  string password = 2;
}


message User {
  string name = 1;
  string email = 2;
  string password = 3;
}

message Users {
  repeated User users = 1;
}

// ================== Chat =====================

message ChatCreateRequest {
  int64 idChat = 1;
  string subject = 2;
  string email = 3;
}

message ChatUpdateRequest {
  int64 idChat = 1;
  string subject = 2;
}

message ChatDeleteRequest {
  int64 idChat = 1;
}

message ChatReadRequest {
  int64 idChat = 1;
  string email = 2;
}


message Chat {
  int64 idChat = 1;
  string subject = 2;
  google.protobuf.Timestamp startDate = 3;
  string email = 4;
}

message Chats {
  repeated Chat chats = 1;
}

// ================== Message =====================

message MessageCreateRequest {
  string content = 1;
  int64 idChat = 2;
}

message MessageReadRequest {
  int64 idMessage = 1;
  int64 idChat = 2;
}

message MessageUpdateRequest {
  int64 idMessage = 1;
  string content = 2;
}

message MessageDeleteRequest {
  int64 idMessage = 1;
}



message Message {
  int64 idMessage = 1;
  string content = 2;
  google.protobuf.Timestamp timestamp = 3;
  int64 idChat = 4;
}

message Messages {
  repeated Message messages = 1;
}

// =======================================
